In [1]:
# Standard data science libraries
import pandas as pd
import numpy as np
from pathlib import Path
from sklearn.preprocessing import MultiLabelBinarizer
from dateutil import parser

import warnings
warnings.filterwarnings('ignore')

In [2]:
from bnb.data_processing.read import read_csvfile
from bnb.data_processing.clean import price_format
from bnb.data_processing.clean import bathrooms_int
from bnb.data_processing.clean import shared_bathroom
from bnb.data_processing.clean import super_host
from bnb.data_processing.clean import beds_int
from bnb.data_processing.clean import bedrooms_int
from bnb.data_processing.clean import bedrooms_int_final
from bnb.data_processing.clean import num_of_stars
from bnb.data_processing.clean import reviews_num
from bnb.data_processing.clean import entire_shared_private_room
from bnb.data_processing.clean import bed_type
from bnb.data_processing.clean import amenities_to_vars
from bnb.data_processing.clean import encode_amenities
from bnb.data_processing.clean import one_hot_encode_bedtype
from bnb.data_processing.clean import amenities_to_vars2
from bnb.data_processing.clean import week_of_year

## Reading Data

In [3]:
Data_dir = Path('C:/Users/KBrig/AirBnB/data')
Data_dir

WindowsPath('C:/Users/KBrig/AirBnB/data')

In [4]:
NewYork_CSV = Data_dir/'AirBnBNY.csv'
Bangkok_CSV = Data_dir/'AirBnbDataBangkok.csv'
Cairo_CSV = Data_dir/'AirBnbDataCairo.csv'

## NewYork

In [5]:
NewYorkData = read_csvfile(NewYork_CSV)
NewYorkData.head(3)


Unnamed: 0       Accuracy_stars  \
0           0    Rating 5 out of 5   
1           1  Rating 4.5 out of 5   
2           2    Rating 5 out of 5   

                                           Amenities  Bathrooms_num  \
0  ['Indoor fireplace', 'Wifi', 'Iron', 'Laptop f...         1 bath   
1  ['Indoor fireplace', 'Wifi', 'Laptop friendly ...  1 shared bath   
2  ['Wifi', 'Essentials', 'Heating', 'Air conditi...  1 shared bath   

  Bedrooms_no Beds_no              CheckIn        CheckIn_stars  \
0   1 bedroom   1 bed  2019-03-21 00:00:00    Rating 5 out of 5   
1   1 bedroom   1 bed  2019-03-21 00:00:00  Rating 4.5 out of 5   
2   1 bedroom   1 bed  2019-03-21 00:00:00    Rating 5 out of 5   

              CheckOut  Cleanliness_stars         ...           Latitude  \
0  2019-03-26 00:00:00  Rating 5 out of 5         ...           40.73321   
1  2019-03-26 00:00:00  Rating 4 out of 5         ...           40.76580   
2  2019-03-26 00:00:00  Rating 5 out of 5         ...           40.75972   

        Location_stars  Longitude  Price  \
0    Rating 5 out of 5  -74.00242   $244   
1    Rating 5 out of 5  -73.95812    $62   
2  Rating 4.5 out of 5  -74.03827    $44   

                                              Review  Reviewers_no  \
0  87 Reviews\nAccuracy\nCommunication\nCleanline...           NaN   
1  5 Reviews\nAccuracy\nCommunication\nCleanlines...           NaN   
2  83 Reviews\nAccuracy\nCommunication\nCleanline...           NaN   

        Sleeping_engagment              Super_host  \
0   Bedroom 1\n1 queen bed  Brian is a Superhost ·   
1  Bedroom 1\n1 single bed    Samet is a Superhost   
2   Bedroom 1\n1 queen bed  Rebecca is a Superhost   

                                               Title          Value_stars  
0  *Luxury 1 Bedrm Private Designer Home & Roofdeck*  Rating 4.5 out of 5  
1        Single Bedroom-Upper East Side-Central Park    Rating 4 out of 5  
2  Quiet room for sleeping - 15 min from Times Sq...    Rating 5 out of 5  

[3 rows x 23 columns]

## Bangkok

In [6]:
BangkokData = read_csvfile(Bangkok_CSV)
BangkokData.head(3)

CheckIn         CheckOut  Has_Weekend  \
0  2019-03-12 0:00  2019-03-17 0:00         True   
1  2019-03-12 0:00  2019-03-17 0:00         True   
2  2019-03-12 0:00  2019-03-17 0:00         True   

                                               Title Price  \
0                Cosy room with 3 mins walk from MRT   $34   
1                                       曼谷市区BTS旁品质公寓   $48   
2  Modern apartment close to city near sky train(...   $27   

                 Super_host  Latitude  Longitude    EntireHomeVsRoom  \
0  Thongjuta is a Superhost  13.78165  100.57453    ENTIRE APARTMENT   
1                       NaN  13.69750  100.60708    ENTIRE APARTMENT   
2       Kiek is a Superhost  13.71252  100.46783  ENTIRE CONDOMINIUM   

  Bedrooms_no                        ...                          \
0   1 bedroom                        ...                           
1   1 bedroom                        ...                           
2   1 bedroom                        ...                           

        Sleeping_engagment                                          Amenities  \
0   Bedroom 1\n1 queen bed  ['Wifi', 'Iron', 'Cable TV', 'Laptop friendly ...   
1  Bedroom 1\n1 double bed  ['Wifi', 'Indoor fireplace', 'Heating', 'Iron'...   
2   Bedroom 1\n1 queen bed  ['Wifi', 'Heating', 'Laptop friendly workspace...   

      Accuracy_stars Communication_stars    Cleanliness_stars  \
0  Rating 5 out of 5   Rating 5 out of 5  Rating 4.5 out of 5   
1                NaN                 NaN                  NaN   
2  Rating 5 out of 5   Rating 5 out of 5    Rating 5 out of 5   

      Location_stars      CheckIn_stars          Value_stars Reviewers_no  \
0  Rating 5 out of 5  Rating 5 out of 5    Rating 5 out of 5          NaN   
1                NaN                NaN                  NaN          NaN   
2  Rating 4 out of 5  Rating 5 out of 5  Rating 4.5 out of 5          NaN   

                                              Review  
0  101 Reviews\nAccuracy\nCommunication\nCleanlin...  
1  No reviews (yet)\nBe one of the first guests t...  
2  4 Reviews\nAccuracy\nCommunication\nCleanlines...  

[3 rows x 22 columns]

## Cairo

In [7]:
CairoData = read_csvfile(Cairo_CSV)
CairoData.head(3)

CheckIn         CheckOut  Has_Weekend  \
0  2019-03-14 0:00  2019-03-19 0:00         True   
1  2019-03-14 0:00  2019-03-19 0:00         True   
2  2019-03-14 0:00  2019-03-19 0:00         True   

                                               Title Price Super_host  \
0                                         Dad Haus!x   $15        NaN   
1                    Apartment in Rehab City, Cairo.   $12        NaN   
2  Sunny and friendly apartment close to the airport   $22        NaN   

   Latitude  Longitude           EntireHomeVsRoom Bedrooms_no  \
0  29.95961   31.25858  Private room in apartment   1 bedroom   
1  30.05760   31.48779           Entire apartment   1 bedroom   
2  30.06085   31.49540  Private room in apartment   1 bedroom   

                         ...                               Sleeping_engagment  \
0                        ...                          Bedroom 1\n2 queen beds   
1                        ...                           Bedroom 1\n1 queen bed   
2                        ...                                              NaN   

                                           Amenities     Accuracy_stars  \
0  ['Wifi', 'Indoor fireplace', 'Dryer', 'Laptop ...                NaN   
1  ['Wifi', 'Laptop friendly workspace', 'Washer'...                NaN   
2  ['Wifi', 'Indoor fireplace', 'Cable TV', 'Lapt...  Rating 4 out of 5   

  Communication_stars    Cleanliness_stars       Location_stars  \
0                 NaN                  NaN                  NaN   
1                 NaN                  NaN                  NaN   
2   Rating 5 out of 5  Rating 3.5 out of 5  Rating 4.5 out of 5   

       CheckIn_stars        Value_stars Reviewers_no  \
0                NaN                NaN          NaN   
1                NaN                NaN          NaN   
2  Rating 5 out of 5  Rating 4 out of 5          NaN   

                                              Review  
0  No reviews (yet)\nBe one of the first guests t...  
1  No reviews (yet)\nThis host has 11 reviews for...  
2  6 Reviews\nAccuracy\nCommunication\nCleanlines...  

[3 rows x 22 columns]

## Data Preprocessing

## NewYork

In [8]:
NewYorkData.shape

(3957, 23)

In [9]:
NewYorkData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3957 entries, 0 to 3956
Data columns (total 23 columns):
Unnamed: 0             3957 non-null int64
Accuracy_stars         3614 non-null object
Amenities              3957 non-null object
Bathrooms_num          3922 non-null object
Bedrooms_no            3348 non-null object
Beds_no                3895 non-null object
CheckIn                3957 non-null object
CheckIn_stars          3614 non-null object
CheckOut               3957 non-null object
Cleanliness_stars      3614 non-null object
Communication_stars    3614 non-null object
EntireHomeVsRoom       3957 non-null object
Has_Weekend            3957 non-null bool
Latitude               3957 non-null float64
Location_stars         3614 non-null object
Longitude              3957 non-null float64
Price                  3917 non-null object
Review                 3956 non-null object
Reviewers_no           0 non-null float64
Sleeping_engagment     3158 non-null object
Super_host      

We have a total of 23 features as shown above: 1 boolean (Has_Weekend), 3 numeric, float64(Latitude, Longitude, Reviewers_num), and the rest are saved as data objects. Let's go through them!! 



So, we need to keep ONLY features that can add information to our analysis. Predictors with many null values can be considered redundant and ignore tham.

### Missing Data

In [10]:
NewYorkData.isnull().sum()

Unnamed: 0                0
Accuracy_stars          343
Amenities                 0
Bathrooms_num            35
Bedrooms_no             609
Beds_no                  62
CheckIn                   0
CheckIn_stars           343
CheckOut                  0
Cleanliness_stars       343
Communication_stars     343
EntireHomeVsRoom          0
Has_Weekend               0
Latitude                  0
Location_stars          343
Longitude                 0
Price                    40
Review                    1
Reviewers_no           3957
Sleeping_engagment      799
Super_host             2010
Title                     0
Value_stars             343
dtype: int64

In [11]:
## Features with large number of missing data
[column for column in NewYorkData.columns if NewYorkData[column].count() / len(NewYorkData) <= 0.7]

['Reviewers_no', 'Super_host']

So "Super_host" can have two possible values: Yes/True or No/False. Samples with null values are actually representing non super_host. On the other hand, having 3957 of null values for the Reviews_no says that there is something wrong with collecting data for this variable. So, it is better to exclude it.

Also, for the reviews' features, we see that we have the same number of null values for the Accuracy,CheckIn,Cleanliness,Communication, Location and Value Stars. This make sense since this number points to the homes with "no reviews yet".  We are going to have this invistigation for all features!

Let's classify the features: numeric, catogerical,etc.


So we have 23 variables six of them are categorical variables reflecting the clients' satisfaction in different aspects: 'Accuracy_stars','CheckIn_stars','Cleanliness_stars','Communication_stars','Location_stars','Value_stars'.

Other categorical variables such as 'EntireHomeVsRoom','Has_Weekend', and 'Sleeping_engagment'.

We also have dates: 'CheckIn','CheckOut'.

We have three numeric features: 'Price','Bathrooms_num', 'Bedrooms_no', 'Beds_no', 'Latitude','Longitude'.

rest:'Amenities', 'Review', 'Title', and 'Super_host'


Let's tidy up!


In [17]:
NewYorkData.drop(['Unnamed: 0','Reviewers_no'],axis=1, inplace=True)

In [50]:
#bin_showings = [len([num for num in num_times_posted  if num == bin]) for bin in range(20)]

In [129]:
#bin_showings

### Does it have a weekend?

We have originally one boolean variable "Has_Weekend". Let's check its levels!

In [16]:
NewYorkData['Has_Weekend'].unique()

array([ True])

So "Has_Weekend" doesn't say more than every time_period is a mix of weekdays and weekends. Thus, we can ignore it!

In [17]:
NewYorkData.drop('Has_Weekend',axis=1, inplace=True)

### Is Super Host?

In [18]:
NewYorkData.Super_host[0:10]

0    Brian is a Superhost ·
1      Samet is a Superhost
2    Rebecca is a Superhost
3      James is a Superhost
4        Cat is a Superhost
5       Rick is a Superhost
6                       NaN
7    Shailah is a Superhost
8                       NaN
9       Obed is a Superhost
Name: Super_host, dtype: object

In [19]:
NewYorkData['Super_host'] = super_host(NewYorkData)

True = 1, False = 0

In [20]:
NewYorkData['Super_host'][0:10]

0    1
1    1
2    1
3    1
4    1
5    1
6    0
7    1
8    0
9    1
Name: Super_host, dtype: int64

## Numeric Data

### Price

In [21]:
print(type(NewYorkData['Price'][0]))
print(NewYorkData['Price'][0:10])

<class 'str'>
0    $244
1     $62
2     $44
3     $78
4     $80
5    $116
6     $96
7    $119
8     $73
9     $63
Name: Price, dtype: object


In [22]:
NewYorkData['Price'] = price_format(NewYorkData)

In [23]:
print(type(NewYorkData['Price'][0]))
print(NewYorkData['Price'][0:10])

<class 'numpy.float64'>
0    244.0
1     62.0
2     44.0
3     78.0
4     80.0
5    116.0
6     96.0
7    119.0
8     73.0
9     63.0
Name: Price, dtype: float64


### Number of Bathrooms

In [24]:
print(type(NewYorkData['Bathrooms_num'][0]))
print(NewYorkData['Bathrooms_num'].unique())

<class 'str'>
['1 bath' '1 shared bath' '2 shared baths' '1.5 shared baths'
 '1.5 private baths' '1 private bath' '2 private baths' nan
 '2.5 shared baths' '2 baths' '2.5 private baths' '3 shared baths'
 '4 shared baths' '0 baths' 'Half-bath' 'Shared half-bath' '1.5 baths'
 '2.5 baths' '3 baths' '0 shared baths' '3.5 shared baths']


In [25]:
NewYorkData['Bathrooms_num_int']= float(0)

In [26]:
NewYorkData['Bathrooms_num_int'] = bathrooms_int(NewYorkData)

In [27]:
for i in range(10):
    print(NewYorkData['Bathrooms_num'][i],NewYorkData['Bathrooms_num_int'][i])

1 bath 1.0
1 shared bath 1.0
1 shared bath 1.0
2 shared baths 2.0
1.5 shared baths 1.5
1 shared bath 1.0
1 shared bath 1.0
1 shared bath 1.0
1 shared bath 1.0
2 shared baths 2.0


### Shared vs. Private Bathroom

In [28]:
NewYorkData['shared_bathroom']= shared_bathroom(NewYorkData)

In [29]:
NewYorkData[['Bathrooms_num','shared_bathroom']][0:10]

Bathrooms_num  shared_bathroom
0            1 bath            False
1     1 shared bath             True
2     1 shared bath             True
3    2 shared baths             True
4  1.5 shared baths             True
5     1 shared bath             True
6     1 shared bath             True
7     1 shared bath             True
8     1 shared bath             True
9    2 shared baths             True

### Number of Bedrooms

In [12]:
print(type(NewYorkData['Bedrooms_no'][0]))

<class 'str'>


In [31]:
NewYorkData['Bedrooms_int'] = None

In [33]:
NewYorkData['Bedrooms_int']= bedrooms_int(NewYorkData)

In [ ]:
for i in range(len(NewYorkData)):
    print(NewYorkData['Bedrooms_no'][i],NewYorkData['Bedrooms_int'][i])

In [35]:
NewYorkData.Bedrooms_int.unique()

array(['1', -1, 'private', 'large', 'tiny', '2', 'Beautiful', '3',
       'Farmers', '4', 'Cozy', 'Private', 'Nice', 'Large', 'size', 'one',
       'Oasis', 'Big', 'big', 'entire', 'One', 'loft', 'master', 'two',
       'Lovely', '5', 'spacious', '"loft"', 'the', 'artistic', 'calm',
       'three', 'Two', 'Midtown', '6', 'Sunny', 'Astoria', 'Comfortable',
       'Peaceful', 'Huge', 'Luxury', 'price', 'Bright', 'our'],
      dtype=object)

In [36]:
NewYorkData[NewYorkData.Bedrooms_int == 'our']['EntireHomeVsRoom']

2668    ROOM IN BOUTIQUE HOTEL
Name: EntireHomeVsRoom, dtype: object

In [37]:
NewYorkData[NewYorkData.Bedrooms_int == 'Astoria']['EntireHomeVsRoom']

2088    PRIVATE ROOM IN APARTMENT
Name: EntireHomeVsRoom, dtype: object

In [38]:
NewYorkData[NewYorkData.Bedrooms_int == 'Oasis']['EntireHomeVsRoom']

331    ENTIRE APARTMENT
Name: EntireHomeVsRoom, dtype: object

In [39]:
oneList = ['private', 'large','tiny','Beautiful', 'Farmers','Cozy', 'Private', 'Nice', 'Large', 'size', 'one',
       'Oasis', 'Big', 'big', 'entire', 'One', 'loft', 'master','Lovely','spacious', 'the', 'artistic', 'calm', 'Midtown','Sunny', 'Astoria', 'Comfortable',
       'Peaceful', 'Huge', 'Luxury', 'price', 'Bright', 'our']
for i in oneList:
    NewYorkData['Bedrooms_int']= NewYorkData['Bedrooms_int'].str.replace(i,'1')
    
twoList =['two','Two']
for i in twoList:
    NewYorkData['Bedrooms_int']= NewYorkData['Bedrooms_int'].str.replace(i,'2')
    

NewYorkData['Bedrooms_int']= NewYorkData['Bedrooms_int'].str.replace('three','3')
NewYorkData['Bedrooms_int']= NewYorkData['Bedrooms_int'].str.replace("loft",'1')

In [40]:
NewYorkData.Bedrooms_int.unique()

array(['1', nan, '2', '3', '4', '5', '"1"', '6'], dtype=object)

In [41]:
NewYorkData['Bedrooms_int']= NewYorkData['Bedrooms_int'].str.replace('"1"','1')

In [42]:
NewYorkData['Bedrooms_int'] = bedrooms_int_final(NewYorkData)
#for i in range(len(NewYorkData)):
#    if(NewYorkData.Bedrooms_int[i]==None):
#        NewYorkData.Bedrooms_int[i]= 0
#    elif(NewYorkData.Bedrooms_int[i] == '"1"'):
#        NewYorkData.Bedrooms_int[i]= 1
#    else:
#        NewYorkData.Bedrooms_int[i]=int(NewYorkData.Bedrooms_int[i])

### Number of Beds

In [45]:
print(type(NewYorkData['Beds_no'][0]))
print(NewYorkData['Beds_no'].unique())

<class 'str'>
['1 bed' '0 beds' 'Obed' '2 beds' 'PRIVATE ROOM IN BED AND BREAKFAST'
 '3 beds' nan '5 beds' '4 beds' '8 beds'
 'SHARED ROOM IN BED AND BREAKFAST' '6 beds' 'ENTIRE BED AND BREAKFAST']


In [46]:
NewYorkData['Beds_int'] = float(0)

In [47]:
NewYorkData['Beds_int'] = beds_int(NewYorkData)

In [48]:
for i in range(10):
    print(NewYorkData['Beds_no'][i],NewYorkData['Beds_int'][i])

1 bed 1.0
1 bed 1.0
1 bed 1.0
1 bed 1.0
1 bed 1.0
1 bed 1.0
1 bed 1.0
0 beds 0.0
1 bed 1.0
Obed 0.0


### Latitude

In [49]:
print(type(NewYorkData['Latitude'][0]))
print(NewYorkData['Latitude'].unique())

<class 'numpy.float64'>
[40.73321 40.7658  40.75972 ... 40.77264 40.76406 40.73354]


### Longitude

In [50]:
print(type(NewYorkData['Longitude'][0]))
print(NewYorkData['Longitude'].unique())

<class 'numpy.float64'>
[-74.00242 -73.95812 -74.03827 ... -74.03762 -74.02755 -73.90769]


Thanks god! we have Latitude and Longitude ready to go!

### Number of Reviews

In [8]:
NewYorkData['Reviewers_no'] = reviews_num(NewYorkData)

In [9]:
NewYorkData['Reviewers_no'].head(10)

0     87
1      5
2     83
3     32
4     28
5     -1
6      8
7    140
8    570
9     19
Name: Reviewers_no, dtype: int64

### Number of Amenities

In [9]:
NewYorkData['Amenities_mod']= None
NewYorkData['Amenities_length']= None

In [10]:
NewYorkData['Amenities_mod']= amenities_to_vars(NewYorkData)

In [13]:
NewYorkData['Amenities_length']= amenities_to_vars2(NewYorkData)

### Bangkok

In [56]:
BangkokData.columns

Index(['CheckIn', 'CheckOut', 'Has_Weekend', 'Title', 'Price', 'Super_host',
       'Latitude', 'Longitude', 'EntireHomeVsRoom', 'Bedrooms_no', 'Beds_no',
       'Bathrooms_num', 'Sleeping_engagment', 'Amenities', 'Accuracy_stars',
       'Communication_stars', 'Cleanliness_stars', 'Location_stars',
       'CheckIn_stars', 'Value_stars', 'Reviewers_no', 'Review'],
      dtype='object')

### Does it have a weekend?

In [57]:
BangkokData['Has_Weekend'].unique()

array([ True])

In [58]:
BangkokData.drop(['Has_Weekend','Reviewers_no'],axis=1, inplace=True)

### Is Super Host?

In [59]:
BangkokData.Super_host[0:10]

0     Thongjuta is a Superhost
1                          NaN
2          Kiek is a Superhost
3                          NaN
4         Patty is a Superhost
5         Nukul is a Superhost
6                          NaN
7                          NaN
8                          NaN
9    Kinifrog is a Superhost ·
Name: Super_host, dtype: object

In [60]:
BangkokData['Super_host'] = super_host(BangkokData)

In [61]:
BangkokData.Super_host[0:10]

0    1
1    0
2    1
3    0
4    1
5    1
6    0
7    0
8    0
9    1
Name: Super_host, dtype: int64

### Price

In [62]:
BangkokData.Price[0:10]

0    $34
1    $48
2    $27
3    $35
4    $34
5    $23
6    $27
7    $51
8    $38
9    $36
Name: Price, dtype: object

In [63]:
BangkokData['Price'] = price_format(BangkokData)

In [64]:
print(type(BangkokData['Price'][0]))
print(BangkokData['Price'][0:10])

<class 'numpy.float64'>
0    34.0
1    48.0
2    27.0
3    35.0
4    34.0
5    23.0
6    27.0
7    51.0
8    38.0
9    36.0
Name: Price, dtype: float64


### Number of Bathrooms

In [65]:
print(type(BangkokData['Bathrooms_num'][0]))
print(BangkokData['Bathrooms_num'].unique())

<class 'str'>
['1 bath' '1 private bath' '2 baths' '1.5 baths' '0 baths' '1 shared bath'
 '2 shared baths' '4 shared baths' '1.5 private baths' '8 shared baths'
 '1.5 shared baths' '3 baths' '4 baths' '2.5 baths' '3.5 baths' nan
 '4.5 baths'
 'Awesome place! Laundry service was available and affordable, a well equipped kitchen was provided, and the bathrooms and rooms were clean.'
 '6 shared baths' '4.5 shared baths']


In [66]:
BangkokData['Bathrooms_num_int']= float(0)

In [ ]:
BangkokData[BangkokData['Bathrooms_num']== 'Awesome place! Laundry service was available and affordable, a well equipped kitchen was provided, and the bathrooms and rooms were clean.']

By looking at the data, I noticed that this sample representing shared room in hostle, and they did not provide any information about the number of bathrooms. For simplicity, I will assume that we have only "1 shared bathroom"! 

In [68]:
for i in range(len(BangkokData)): 
    if(BangkokData['Bathrooms_num'][i]== 'Awesome place! Laundry service was available and affordable, a well equipped kitchen was provided, and the bathrooms and rooms were clean.'):
        BangkokData['Bathrooms_num'][i]= '1 shared bath'
        
    

In [69]:
print(BangkokData['Bathrooms_num'].unique())

['1 bath' '1 private bath' '2 baths' '1.5 baths' '0 baths' '1 shared bath'
 '2 shared baths' '4 shared baths' '1.5 private baths' '8 shared baths'
 '1.5 shared baths' '3 baths' '4 baths' '2.5 baths' '3.5 baths' nan
 '4.5 baths' '6 shared baths' '4.5 shared baths']


In [70]:
BangkokData['Bathrooms_num_int'] = bathrooms_int(BangkokData)

In [71]:
for i in range(10):
    print(BangkokData['Bathrooms_num'][i],BangkokData['Bathrooms_num_int'][i])

1 bath 1.0
1 bath 1.0
1 bath 1.0
1 bath 1.0
1 bath 1.0
1 bath 1.0
1 private bath 1.0
1 bath 1.0
1 bath 1.0
1 bath 1.0


### Shared vs. Private Bathroom

In [72]:
BangkokData['shared_bathroom']= shared_bathroom(BangkokData)

In [214]:
BangkokData[['Bathrooms_num','shared_bathroom']][0:10]

Bathrooms_num  shared_bathroom
0          1 bath            False
1          1 bath            False
2          1 bath            False
3          1 bath            False
4          1 bath            False
5          1 bath            False
6  1 private bath            False
7          1 bath            False
8          1 bath            False
9          1 bath            False

In [215]:
BangkokData.shared_bathroom  = [int(i == True) for i in BangkokData.shared_bathroom]


### Number of Bedrooms 

In [13]:
print(type(BangkokData['Bedrooms_no'][0]))

<class 'str'>


In [75]:
BangkokData.Bedrooms_no = BangkokData.Bedrooms_no.str.replace('360°view2bedrooms','360° view 2 bedrooms')
BangkokData.Bedrooms_no = BangkokData.Bedrooms_no.str.replace('one-bedroom','1 bedroom')
BangkokData.Bedrooms_no = BangkokData.Bedrooms_no.str.replace('1bedroom','1 bedroom')
BangkokData.Bedrooms_no = BangkokData.Bedrooms_no.str.replace('two bedroom','2 bedrooms')
BangkokData.Bedrooms_no = BangkokData.Bedrooms_no.str.replace('The Zen bedroom','5 bedrooms')
BangkokData.Bedrooms_no = BangkokData.Bedrooms_no.str.replace('One bedroom','1 bedroom')
BangkokData.Bedrooms_no = BangkokData.Bedrooms_no.str.replace('one bedroom','1 bedroom')
BangkokData.Bedrooms_no = BangkokData.Bedrooms_no.str.replace('2bedrooms','2 bedrooms')
BangkokData.Bedrooms_no = BangkokData.Bedrooms_no.str.replace('4-bedrooms','4 bedrooms')
BangkokData.Bedrooms_no = BangkokData.Bedrooms_no.str.replace('3-bedrooms','3 bedrooms')


In [76]:
BangkokData['Bedrooms_int']= None 

In [77]:
BangkokData['Bedrooms_int'] = bedrooms_int(BangkokData)

In [78]:
BangkokData['Bedrooms_int'].unique()

array(['1', -1, 'king', 'the', 'Single', '2', 'Big', 'King', 'pool-Comfy',
       'size', '5', 'BTS/1', 'from', '3', '4', '6', '☼Sky'], dtype=object)

In [79]:
oneList = ['king', 'the', 'Single', 'Big', 'King','pool-Comfy', 'size', 'BTS/1', 'from', '☼Sky']
for i in oneList:
    BangkokData['Bedrooms_int']= BangkokData['Bedrooms_int'].replace(i,'1')


In [80]:
BangkokData['Bedrooms_int'].unique()

array(['1', -1, '2', '5', '3', '4', '6'], dtype=object)

In [81]:
BangkokData['Bedrooms_int']= bedrooms_int_final(BangkokData)

In [84]:
BangkokData['Bedrooms_int'].head(10)

0     1
1     1
2     1
3     1
4    -1
5     1
6    -1
7     1
8     1
9     1
Name: Bedrooms_int, dtype: object

### Number of Beds

In [85]:
print(type(BangkokData['Beds_no'][0]))
print(BangkokData['Beds_no'].unique())

<class 'str'>
['1 bed' '2 beds' '3 beds' nan '4 beds' '0 beds'
 'PRIVATE ROOM IN BED AND BREAKFAST' '5 beds' '6 beds' '7 beds' '8 beds'
 'Private room in bed and breakfast']


In [86]:
BangkokData['Beds_int'] = float(0)

In [87]:
BangkokData['Beds_int']= beds_int(BangkokData)

In [88]:
for i in range(10):
    print(BangkokData['Beds_no'][i],BangkokData['Beds_int'][i])

1 bed 1.0
1 bed 1.0
2 beds 2.0
1 bed 1.0
3 beds 3.0
3 beds 3.0
1 bed 1.0
1 bed 1.0
1 bed 1.0
1 bed 1.0


### Latitude

In [89]:
print(type(BangkokData['Latitude'][0]))

<class 'numpy.float64'>


### Longitude

In [90]:
print(type(BangkokData['Longitude'][0]))

<class 'numpy.float64'>


### Number of Reviews

In [10]:
BangkokData['Reviewers_no'] = reviews_num(BangkokData)

In [11]:
BangkokData['Reviewers_no'].head(10)

0    101
1     -1
2      4
3      3
4     -1
5     23
6      1
7     10
8      2
9     82
Name: Reviewers_no, dtype: int64

### Number of Amenities

In [16]:
BangkokData['Amenities_mod']= None
BangkokData['Amenities_length']= None

In [17]:
BangkokData['Amenities_mod']= amenities_to_vars(BangkokData)

In [18]:
BangkokData['Amenities_length']= amenities_to_vars2(BangkokData)

# Cairo

In [91]:
CairoData.columns

Index(['CheckIn', 'CheckOut', 'Has_Weekend', 'Title', 'Price', 'Super_host',
       'Latitude', 'Longitude', 'EntireHomeVsRoom', 'Bedrooms_no', 'Beds_no',
       'Bathrooms_num', 'Sleeping_engagment', 'Amenities', 'Accuracy_stars',
       'Communication_stars', 'Cleanliness_stars', 'Location_stars',
       'CheckIn_stars', 'Value_stars', 'Reviewers_no', 'Review'],
      dtype='object')

### Does it have a weekend?

In [92]:
CairoData['Has_Weekend'].unique()

array([ True])

In [93]:
CairoData.drop(['Has_Weekend','Reviewers_no'],axis=1, inplace=True)

In [94]:
CairoData.columns

Index(['CheckIn', 'CheckOut', 'Title', 'Price', 'Super_host', 'Latitude',
       'Longitude', 'EntireHomeVsRoom', 'Bedrooms_no', 'Beds_no',
       'Bathrooms_num', 'Sleeping_engagment', 'Amenities', 'Accuracy_stars',
       'Communication_stars', 'Cleanliness_stars', 'Location_stars',
       'CheckIn_stars', 'Value_stars', 'Review'],
      dtype='object')

### Is Super Host?

In [95]:
CairoData.Super_host[0:10]

0                       NaN
1                       NaN
2                       NaN
3                       NaN
4                       NaN
5                       NaN
6                       NaN
7     Bassem is a Superhost
8                       NaN
9    Mohamed is a Superhost
Name: Super_host, dtype: object

In [96]:
CairoData['Super_host'] = super_host(CairoData)

In [97]:
CairoData.Super_host[0:10]

0    0
1    0
2    0
3    0
4    0
5    0
6    0
7    1
8    0
9    1
Name: Super_host, dtype: int64

### Price

In [98]:
CairoData['Price'] = price_format(CairoData)

In [99]:
print(type(CairoData['Price'][0]))
print(CairoData['Price'][0:10])

<class 'numpy.float64'>
0    15.0
1    12.0
2    22.0
3    19.0
4    23.0
5    25.0
6    40.0
7    37.0
8    65.0
9    11.0
Name: Price, dtype: float64


### Number of Bathrooms

In [100]:
print(type(CairoData['Bathrooms_num'][0]))
print(CairoData['Bathrooms_num'].unique())

<class 'str'>
['2 shared baths' '1.5 baths' '1 shared bath' '1.5 private baths'
 '1 private bath' '1 bath' '2 baths' '3 baths' '12 private baths'
 '8 private baths' '1.5 shared baths' '2.5 baths' nan '4 baths'
 '2 private baths' '0 shared baths' '3 private baths' '3 shared baths'
 '10 private baths' '5 private baths' '6 baths' '4 private baths'
 '3.5 baths' '4 shared baths' '25 private baths' '20 private baths']


In [101]:
CairoData['Bathrooms_num_int']= float(0)

In [102]:
CairoData['Bathrooms_num_int']= bathrooms_int(CairoData) 

In [103]:
for i in range(10):
    print(CairoData['Bathrooms_num'][i],CairoData['Bathrooms_num_int'][i])

2 shared baths 2.0
1.5 baths 1.5
1 shared bath 1.0
1.5 private baths 1.5
1 shared bath 1.0
1 private bath 1.0
1 private bath 1.0
1.5 baths 1.5
1 bath 1.0
1 shared bath 1.0


### Shared vs. Private Bathrooms

In [104]:
CairoData['shared_bathroom']= shared_bathroom(CairoData)

In [105]:
CairoData[['Bathrooms_num','shared_bathroom']][0:10]

Bathrooms_num  shared_bathroom
0     2 shared baths             True
1          1.5 baths            False
2      1 shared bath             True
3  1.5 private baths            False
4      1 shared bath             True
5     1 private bath            False
6     1 private bath            False
7          1.5 baths            False
8             1 bath            False
9      1 shared bath             True

In [217]:
CairoData.shared_bathroom  = [int(i == True) for i in CairoData.shared_bathroom]


### Number of Bedrooms

In [14]:
print(type(CairoData['Bedrooms_no'][0]))

<class 'str'>


In [107]:
CairoData['Bedrooms_no'] = CairoData['Bedrooms_no'].str.replace('2 Single bedroom','2 bedroom')
CairoData['Bedrooms_no'] = CairoData['Bedrooms_no'].str.replace('3bedroom','3 bedroom')
CairoData['Bedrooms_no'] = CairoData['Bedrooms_no'].str.replace('2bedroom','2 bedroom')
CairoData['Bedrooms_no'] = CairoData['Bedrooms_no'].str.replace('2-bedroom','2 bedroom')
CairoData['Bedrooms_no'] = CairoData['Bedrooms_no'].str.replace('BOTH (2) comfortable DBL bedrooms','2 bedroom')
CairoData['Bedrooms_no'] = CairoData['Bedrooms_no'].str.replace('one bedroom','1 bedroom')
CairoData['Bedrooms_no'] = CairoData['Bedrooms_no'].str.replace('One bedroom','1 bedroom')
CairoData['Bedrooms_no'] = CairoData['Bedrooms_no'].str.replace('Three bedroom','3 bedroom')
CairoData['Bedrooms_no'] = CairoData['Bedrooms_no'].str.replace('Two bedroom','3 bedroom')


In [108]:
CairoData['Bedrooms_int']= None 

In [109]:
CairoData['Bedrooms_int']= bedrooms_int(CairoData)

In [110]:
CairoData['Bedrooms_int'].unique()

array(['1', '2', 'cosy', -1, '20', '3', 'Double', 'queen', 'Apartment(3',
       '4', 'Private', '10', 'Apartment(2', '6', '15', '☀☼Sunny', 'Guest',
       'nice', '5', '13', '9', 'DBL', '25', 'Cozy', 'Master'],
      dtype=object)

In [111]:
oneList = ['cosy','Double', 'queen','Private','☀☼Sunny','Guest','nice','DBL','Cozy','Master']
for i in oneList:
    CairoData['Bedrooms_int']= CairoData['Bedrooms_int'].replace(i,'1')
    
CairoData['Bedrooms_int']= CairoData['Bedrooms_int'].replace('Apartment(3','3')
CairoData['Bedrooms_int']= CairoData['Bedrooms_int'].replace('Apartment(2','2')

In [112]:
CairoData['Bedrooms_int'].unique()

array(['1', '2', -1, '20', '3', '4', '10', '6', '15', '5', '13', '9',
       '25'], dtype=object)

In [113]:
CairoData['Bedrooms_int'] = bedrooms_int_final(CairoData)

In [114]:
CairoData['Bedrooms_int'].unique()

array(['1', '2', -1, '20', '3', '4', '10', '6', '15', '5', '13', '9',
       '25'], dtype=object)

### Number of Beds

In [115]:
print(type(CairoData['Beds_no'][0]))
print(CairoData['Beds_no'].unique())

<class 'str'>
['2 beds' '1 bed' '4 beds' nan 'Private room in bed and breakfast'
 '3 beds' '26 beds' '6 beds' '5 beds' '8 beds'
 'Stylish brand new fully furnished studio apartment with queen size bed and open plan kitchen. kitchen is fully equipped with all cutlery and crockery, washing machine, cooker, microwave, kettle, toaster. bed linen and towels are provided.'
 '16 beds' '50 beds' '37 beds' '7 beds' '10 beds' '18 beds' '30 beds']


In [ ]:
CairoData[CairoData['Beds_no']== 'Stylish brand new fully furnished studio apartment with queen size bed and open plan kitchen. kitchen is fully equipped with all cutlery and crockery, washing machine, cooker, microwave, kettle, toaster. bed linen and towels are provided.'][0]

In [117]:
for i in range(len(BangkokData)): 
    if(CairoData['Beds_no'][i]== 'Stylish brand new fully furnished studio apartment with queen size bed and open plan kitchen. kitchen is fully equipped with all cutlery and crockery, washing machine, cooker, microwave, kettle, toaster. bed linen and towels are provided.'):
        CairoData['Beds_no'][i]= '1 bed'

In [118]:
CairoData['Beds_int']= float(0)

In [119]:
CairoData['Beds_int']= beds_int(CairoData)

In [120]:
for i in range(10):
    print(CairoData['Beds_no'][i],CairoData['Beds_int'][i])

2 beds 2.0
1 bed 1.0
1 bed 1.0
2 beds 2.0
1 bed 1.0
1 bed 1.0
1 bed 1.0
2 beds 2.0
1 bed 1.0
1 bed 1.0


### Latitude

In [121]:
print(type(CairoData['Latitude'][0]))

<class 'numpy.float64'>


### Longitude

In [122]:
print(type(CairoData['Longitude'][0]))

<class 'numpy.float64'>


### Number of Reviews

In [12]:
CairoData['Reviewers_no'] = reviews_num(CairoData)

In [13]:
CairoData['Reviewers_no'].head(10)

0    -1
1    -1
2     6
3     4
4     2
5     8
6    19
7    -1
8    -1
9    -1
Name: Reviewers_no, dtype: int64

### Number of Amenities

In [26]:
CairoData['Amenities_mod']= None
CairoData['Amenities_length']= None

In [27]:
CairoData['Amenities_mod']= amenities_to_vars(CairoData)

In [28]:
CairoData['Amenities_length']= amenities_to_vars2(CairoData)

## Categorical Data

### Review Stars

In [123]:
stars_list =['Accuracy_stars','CheckIn_stars','Cleanliness_stars','Communication_stars','Location_stars','Value_stars']

### NewYork

In [124]:
for i in stars_list:
    print(NewYorkData[i].unique())


['Rating 5 out of 5' 'Rating 4.5 out of 5' nan 'Rating 4 out of 5'
 'Rating 2.5 out of 5' 'Rating 3.5 out of 5']
['Rating 5 out of 5' 'Rating 4.5 out of 5' nan 'Rating 4 out of 5'
 'Rating 3.5 out of 5' 'Rating 2 out of 5']
['Rating 5 out of 5' 'Rating 4 out of 5' nan 'Rating 3.5 out of 5'
 'Rating 4.5 out of 5' 'Rating 3 out of 5' 'Rating 2.5 out of 5']
['Rating 5 out of 5' 'Rating 4.5 out of 5' nan 'Rating 4 out of 5'
 'Rating 3.5 out of 5' 'Rating 2 out of 5']
['Rating 5 out of 5' 'Rating 4.5 out of 5' nan 'Rating 4 out of 5'
 'Rating 3.5 out of 5']
['Rating 4.5 out of 5' 'Rating 4 out of 5' 'Rating 5 out of 5' nan
 'Rating 2 out of 5' 'Rating 3.5 out of 5']


In [ ]:
num_of_stars(NewYorkData)

In [126]:
for i in stars_list:
    print(NewYorkData[i].unique())


['5' '4.5' '0' '4' '2.5' '3.5']
['5' '4.5' '0' '4' '3.5' '2']
['5' '4' '0' '3.5' '4.5' '3' '2.5']
['5' '4.5' '0' '4' '3.5' '2']
['5' '4.5' '0' '4' '3.5']
['4.5' '4' '5' '0' '2' '3.5']


### Entire Apartment vs. Shared Room vs. Private Room

In [127]:
NewYorkData.EntireHomeVsRoom.unique()

array(['ENTIRE APARTMENT', 'PRIVATE ROOM IN APARTMENT',
       'PRIVATE ROOM IN HOUSE', 'PRIVATE ROOM IN LOFT',
       'PRIVATE ROOM IN TOWNHOUSE', 'PRIVATE ROOM IN BED AND BREAKFAST',
       'SHARED ROOM IN APARTMENT', 'PRIVATE ROOM IN CONDOMINIUM',
       'PRIVATE ROOM IN SERVICED APARTMENT',
       'PRIVATE ROOM IN GUEST SUITE', 'PRIVATE ROOM IN HOSTEL',
       'ENTIRE LOFT', 'ENTIRE TOWNHOUSE', 'SHARED ROOM IN LOFT',
       'ENTIRE HOUSE', 'SHARED ROOM IN APARTHOTEL',
       'ENTIRE SERVICED APARTMENT', 'ENTIRE CONDOMINIUM', 'PRIVATE ROOM',
       'ROOM IN BOUTIQUE HOTEL', 'SHARED ROOM IN BED AND BREAKFAST',
       'PRIVATE ROOM IN GUESTHOUSE', 'ENTIRE GUEST SUITE',
       'PRIVATE ROOM IN RESORT', 'ENTIRE GUESTHOUSE',
       'SHARED ROOM IN HOUSE', 'PRIVATE ROOM IN TINY HOUSE',
       'ENTIRE BED AND BREAKFAST'], dtype=object)

In [128]:
## Entire apartment = 0
## Shared room = 1
## Private room = 2
NewYorkData['shared_room']= int(0)

In [130]:
NewYorkData['EntireHomeVsRoom'].head(10)

0             ENTIRE APARTMENT
1    PRIVATE ROOM IN APARTMENT
2        PRIVATE ROOM IN HOUSE
3    PRIVATE ROOM IN APARTMENT
4    PRIVATE ROOM IN APARTMENT
5    PRIVATE ROOM IN APARTMENT
6         PRIVATE ROOM IN LOFT
7    PRIVATE ROOM IN APARTMENT
8    PRIVATE ROOM IN APARTMENT
9    PRIVATE ROOM IN TOWNHOUSE
Name: EntireHomeVsRoom, dtype: object

In [131]:
NewYorkData['shared_room']= entire_shared_private_room(NewYorkData)

In [132]:
for i in range(20):
    print(NewYorkData.EntireHomeVsRoom[i],NewYorkData['shared_room'][i])

ENTIRE APARTMENT 0
PRIVATE ROOM IN APARTMENT 2
PRIVATE ROOM IN HOUSE 2
PRIVATE ROOM IN APARTMENT 2
PRIVATE ROOM IN APARTMENT 2
PRIVATE ROOM IN APARTMENT 2
PRIVATE ROOM IN LOFT 2
PRIVATE ROOM IN APARTMENT 2
PRIVATE ROOM IN APARTMENT 2
PRIVATE ROOM IN TOWNHOUSE 2
ENTIRE APARTMENT 0
PRIVATE ROOM IN APARTMENT 2
PRIVATE ROOM IN APARTMENT 2
PRIVATE ROOM IN HOUSE 2
PRIVATE ROOM IN APARTMENT 2
PRIVATE ROOM IN BED AND BREAKFAST 2
ENTIRE APARTMENT 0
SHARED ROOM IN APARTMENT 1
PRIVATE ROOM IN APARTMENT 2
PRIVATE ROOM IN APARTMENT 2


### Sleeping Engagment

In [133]:
for i in NewYorkData.Sleeping_engagment:
    if(str(i)!='nan'):
        if(i.split(' ')[2]=='small'):
            print(i)

Bedroom 1
1 small double bed
Bedroom 1
1 small double bed
Bedroom 1
1 small double bed
Bedroom 1
1 small double bed
Bedroom 1
1 small double bed
Bedroom 1
1 small double bed


In [134]:
NewYorkData.Sleeping_engagment = NewYorkData.Sleeping_engagment.str.replace('small double','double')

In [135]:
NewYorkData['BedType']= None
NewYorkData['BedType'] = bed_type(NewYorkData)
    

In [136]:
NewYorkData['BedType'].unique()

array(['queen', 'single', None, 'king', 'double', 'bunk', 'couch', 'sofa',
       'couches', 'couch,', 'crib', 'air', 'floor'], dtype=object)

In [140]:
NewYorkData = one_hot_encode_bedtype(NewYorkData)

In [142]:
NewYorkData.columns

Index(['Accuracy_stars', 'Amenities', 'Bathrooms_num', 'Bedrooms_no',
       'Beds_no', 'CheckIn', 'CheckIn_stars', 'CheckOut', 'Cleanliness_stars',
       'Communication_stars', 'EntireHomeVsRoom', 'Latitude', 'Location_stars',
       'Longitude', 'Price', 'Review', 'Sleeping_engagment', 'Super_host',
       'Title', 'Value_stars', 'Bathrooms_num_int', 'shared_bathroom',
       'Bedrooms_int', 'Beds_int', 'shared_room', 'BedType', 'air', 'bunk',
       'couch', 'couch,', 'couches', 'crib', 'double', 'floor', 'king',
       'queen', 'single', 'sofa'],
      dtype='object')

In [148]:
NewYorkData.drop(['BedType','couch,', 'couches'],axis=1, inplace=True)

### Amenities

In [19]:
NewYorkData['Amenities_mod']= None
NewYorkData['Amenities_length']= None

In [20]:
NewYorkData = amenities_to_vars(NewYorkData)

In [25]:
NewYorkData[0]

['Indoor fireplace',
 'Wifi',
 'Iron',
 'Laptop friendly workspace',
 'TV',
 'Essentials',
 'Heating',
 'Air conditioning',
 'Hot water',
 'Bathtub',
 'Fireplace guards',
 'Kitchen',
 'Cooking basics',
 'Dishes and silverware',
 'Dishwasher',
 'Microwave',
 'Refrigerator',
 'Oven',
 'Stove',
 'Lockbox',
 'Luggage dropoff allowed',
 'Long term stays allowed',
 'Hangers',
 'Hair dryer',
 'Shampoo',
 'Patio or balcony',
 'Carbon monoxide detector',
 'Smoke detector']

In [152]:
len(NewYorkData.Amenities_mod[0])

28

In [153]:
NewYorkData.Amenities_mod[0]

['Indoor fireplace',
 'Wifi',
 'Iron',
 'Laptop friendly workspace',
 'TV',
 'Essentials',
 'Heating',
 'Air conditioning',
 'Hot water',
 'Bathtub',
 'Fireplace guards',
 'Kitchen',
 'Cooking basics',
 'Dishes and silverware',
 'Dishwasher',
 'Microwave',
 'Refrigerator',
 'Oven',
 'Stove',
 'Lockbox',
 'Luggage dropoff allowed',
 'Long term stays allowed',
 'Hangers',
 'Hair dryer',
 'Shampoo',
 'Patio or balcony',
 'Carbon monoxide detector',
 'Smoke detector']

In [193]:
NewYorkData=encode_amenities(NewYorkData)

In [194]:
NewYorkData.columns

Index(['Accuracy_stars', 'Amenities', 'Bathrooms_num', 'Bedrooms_no',
       'Beds_no', 'CheckIn', 'CheckIn_stars', 'CheckOut', 'Cleanliness_stars',
       'Communication_stars',
       ...
       'Smart lock', 'Smoke detector', 'Stair gates', 'Stove', 'TV',
       'Table corner guards', 'Washer', 'Waterfront', 'Wifi', 'Window guards'],
      dtype='object', length=115)

In [198]:
NewYorkData.drop(['Amenities'],axis=1, inplace = True)

In [219]:
NewYorkData.drop(['Title','EntireHomeVsRoom','Bedrooms_no','Beds_no','Bathrooms_num','Sleeping_engagment','Review'],axis=1 ,inplace = True)

### Bangkok

### Review Stars

In [ ]:
for i in stars_list:
    print(BangkokData[i])

In [ ]:
num_of_stars(BangkokData)

In [156]:
for i in stars_list:
    print(BangkokData[i].unique())

['5' '0' '4.5' '4' '3.5']
['5' '0' '4.5' '4' '3.5' '3']
['4.5' '0' '5' '4' '3.5']
['5' '0' '4.5' '4' '3.5']
['5' '0' '4' '4.5' '3.5' '3']
['5' '0' '4.5' '4' '2.5' '3.5']


### Entire Apartment vs. Shared Room vs. Private Room

In [157]:
BangkokData.EntireHomeVsRoom.unique()

array(['ENTIRE APARTMENT', 'ENTIRE CONDOMINIUM',
       'PRIVATE ROOM IN APARTMENT', 'ENTIRE SERVICED APARTMENT',
       'ROOM IN BOUTIQUE HOTEL', 'ENTIRE LOFT',
       'PRIVATE ROOM IN CONDOMINIUM', 'PRIVATE ROOM IN VILLA',
       'PRIVATE ROOM IN SERVICED APARTMENT', 'PRIVATE ROOM IN TOWNHOUSE',
       'PRIVATE ROOM IN LOFT', 'ENTIRE PLACE', 'ENTIRE TOWNHOUSE',
       'ROOM IN HOTEL', 'ENTIRE GUEST SUITE', 'PRIVATE ROOM IN HOSTEL',
       'ENTIRE HOUSE', 'PRIVATE ROOM IN GUESTHOUSE',
       'PRIVATE ROOM IN BED AND BREAKFAST', 'ENTIRE VILLA',
       'ENTIRE GUESTHOUSE', 'PRIVATE ROOM IN HOUSE', 'ROOM IN APARTHOTEL',
       'SHARED ROOM IN HOSTEL', 'ENTIRE HOSTEL', 'TINY HOUSE',
       'ENTIRE COTTAGE', nan], dtype=object)

In [158]:
## Entire apartment = 0
## Shared room = 1
## Private room = 2
BangkokData['shared_room']= int(0)

In [159]:
BangkokData['shared_room']=entire_shared_private_room(BangkokData)


In [160]:
for i in range(30):
    print(BangkokData.EntireHomeVsRoom[i],BangkokData['shared_room'][i])

ENTIRE APARTMENT 0
ENTIRE APARTMENT 0
ENTIRE CONDOMINIUM 0
ENTIRE CONDOMINIUM 0
ENTIRE APARTMENT 0
ENTIRE CONDOMINIUM 0
PRIVATE ROOM IN APARTMENT 2
ENTIRE APARTMENT 0
ENTIRE CONDOMINIUM 0
ENTIRE APARTMENT 0
ENTIRE APARTMENT 0
ENTIRE SERVICED APARTMENT 0
ENTIRE CONDOMINIUM 0
ROOM IN BOUTIQUE HOTEL 2
ENTIRE APARTMENT 0
ENTIRE APARTMENT 0
ENTIRE APARTMENT 0
ENTIRE CONDOMINIUM 0
ENTIRE APARTMENT 0
ENTIRE CONDOMINIUM 0
ENTIRE APARTMENT 0
ENTIRE CONDOMINIUM 0
ENTIRE APARTMENT 0
ENTIRE APARTMENT 0
ENTIRE CONDOMINIUM 0
ENTIRE APARTMENT 0
ENTIRE CONDOMINIUM 0
ENTIRE APARTMENT 0
ENTIRE LOFT 0
ENTIRE APARTMENT 0


### Bed Type

In [161]:
for i in BangkokData.Sleeping_engagment:
    if(str(i)!='nan'):
        if(i.split(' ')[2]=='small'):
            print(i)

In [162]:
BangkokData.Sleeping_engagment = BangkokData.Sleeping_engagment.str.replace('small double','double')

In [163]:
BangkokData['BedType']= None
BangkokData['BedType']= bed_type(BangkokData)

In [164]:
BangkokData['BedType'].unique()

array(['queen', 'double', 'king', None, 'sofa', 'floor', 'single',
       'couch', 'air', 'couch,', 'bunk'], dtype=object)

In [165]:
BangkokData = one_hot_encode_bedtype(BangkokData)

In [167]:
BangkokData.drop(['BedType','couch,'],axis=1, inplace=True)

### Amenities

In [ ]:
BangkokData.Amenities.unique()

In [169]:
print(BangkokData.Amenities[0])

['Wifi', 'Iron', 'Cable TV', 'Laptop friendly workspace', 'Dryer', 'Washer', 'Essentials', 'Air conditioning', 'TV', 'Hot water', 'Elevator', 'Gym', 'Pool', 'Kitchen', 'Coffee maker', 'Cooking basics', 'Dishes and silverware', 'Microwave', 'Refrigerator', 'Stove', 'Long term stays allowed', 'Hair dryer', 'Hangers', 'Shampoo', 'Smoke detector']


In [170]:
BangkokData['Amenities_mod'] = None

In [171]:
BangkokData['Amenities_mod']= amenities_to_vars(BangkokData)

In [172]:
BangkokData['Amenities_mod'][0:3]

0    [Wifi, Iron, Cable TV, Laptop friendly workspa...
1    [Wifi, Indoor fireplace, Heating, Iron, Laptop...
2    [Wifi, Heating, Laptop friendly workspace, Dry...
Name: Amenities_mod, dtype: object

In [199]:
BangkokData = encode_amenities(BangkokData)

In [202]:
BangkokData.drop(['Amenities'],axis =1, inplace = True)

In [220]:
BangkokData.drop(['Title','EntireHomeVsRoom','Bedrooms_no','Beds_no','Bathrooms_num','Sleeping_engagment','Review'],axis=1 ,inplace = True)

### Cairo

### Review Stars

In [ ]:
num_of_stars(CairoData)

In [176]:
for i in stars_list:
    print(CairoData[i].unique())


['0' '4' '5' '4.5' '3.5' '3']
['0' '5' '4.5' '4' '3.5']
['0' '3.5' '5' '4.5' '4' '3']
['0' '5' '4.5' '4' '3.5']
['0' '4.5' '5' '4' '3.5']
['0' '4' '5' '4.5' '3.5' '3']


### Entire Apartment vs. Shared Room vs. Private Room

In [177]:
CairoData.EntireHomeVsRoom.unique()

array(['Private room in apartment', 'Entire apartment',
       'Private room in condominium', 'Room in hotel',
       'Private room in hostel', 'Entire serviced apartment',
       'Entire condominium', 'Private room in bed and breakfast',
       'Entire house', 'Room in boutique hotel', 'Private room in house',
       'Private room in guesthouse', 'Private room in serviced apartment',
       'Shared room in hostel', 'Private room in tiny house',
       'Tiny house', 'Private room in villa', 'Shared room in apartment',
       'Entire villa', 'Earth house', 'Private room in guest suite',
       'Shared room in villa', 'Entire guest suite',
       'Private room in loft', 'Entire loft', 'Entire townhouse',
       'Private room in boat', 'Entire chalet', 'Shared room in house'],
      dtype=object)

In [178]:
CairoData['shared_room']= int(0)

In [179]:
CairoData['shared_room'] = entire_shared_private_room(CairoData)

In [180]:
for i in range(20):
    print(CairoData.EntireHomeVsRoom[i],CairoData['shared_room'][i])

Private room in apartment 2
Entire apartment 0
Private room in apartment 2
Private room in condominium 2
Private room in apartment 2
Private room in apartment 2
Room in hotel 2
Entire apartment 0
Entire apartment 0
Private room in hostel 2
Private room in apartment 2
Entire serviced apartment 0
Private room in apartment 2
Entire serviced apartment 0
Room in hotel 2
Private room in apartment 2
Entire condominium 0
Private room in bed and breakfast 2
Entire apartment 0
Private room in apartment 2


### Bed Type

In [181]:
for i in CairoData.Sleeping_engagment:
    if(str(i)!='nan'):
        if(i.split(' ')[2]=='small'):
            print(i)

Bedroom 1
2 small double beds
Bedroom 1
1 small double bed
Bedroom 1
1 small double bed
Bedroom 1
1 small double bed
Bedroom 1
1 small double bed
Bedroom 1
1 small double bed
Bedroom 1
1 small double bed
Bedroom 1
1 small double bed
Bedroom 1
1 small double bed
Bedroom 1
1 small double bed
Bedroom 1
1 small double bed


In [182]:
CairoData.Sleeping_engagment = CairoData.Sleeping_engagment.str.replace('small double','double')

In [183]:
CairoData['BedType']= None
CairoData['BedType']= bed_type(CairoData)

In [184]:
CairoData['BedType'].unique()

array(['queen', None, 'single', 'double', 'sofa', 'king', 'floor',
       'couch', 'couches', 'couch,', 'air', 'bunk'], dtype=object)

In [186]:
CairoData = one_hot_encode_bedtype(CairoData)

In [187]:
CairoData.drop(['couches', 'couch,','BedType'],axis=1, inplace=True)

### Amenities

In [ ]:
CairoData.Amenities.unique()

In [189]:
CairoData.Amenities[0]

"['Wifi', 'Indoor fireplace', 'Dryer', 'Laptop friendly workspace', 'Iron', 'Heating', 'Washer', 'TV', 'Air conditioning', 'Essentials', 'Elevator', 'Free parking on premises', 'Kitchen', 'Breakfast', 'Private living room', 'Hair dryer', 'Shampoo', 'Hangers', 'Lock on bedroom door', 'Fire extinguisher', 'Carbon monoxide detector', 'Smoke detector', 'First aid kit']"

In [190]:
CairoData['Amenities_mod'] = None

In [191]:
CairoData['Amenities_mod'] = amenities_to_vars(CairoData)

In [192]:
CairoData.Amenities_mod[0]

['Wifi',
 'Indoor fireplace',
 'Dryer',
 'Laptop friendly workspace',
 'Iron',
 'Heating',
 'Washer',
 'TV',
 'Air conditioning',
 'Essentials',
 'Elevator',
 'Free parking on premises',
 'Kitchen',
 'Breakfast',
 'Private living room',
 'Hair dryer',
 'Shampoo',
 'Hangers',
 'Lock on bedroom door',
 'Fire extinguisher',
 'Carbon monoxide detector',
 'Smoke detector',
 'First aid kit']

In [203]:
CairoData= encode_amenities(CairoData)

In [205]:
CairoData.drop(['Amenities'],axis=1 ,inplace = True)

In [221]:
CairoData.drop(['Title','EntireHomeVsRoom','Bedrooms_no','Beds_no','Bathrooms_num','Sleeping_engagment','Review'],axis=1 ,inplace = True)

### NewYork

In [222]:
print(len(NewYorkData.columns))
NewYorkData.columns

107


Index(['Accuracy_stars', 'CheckIn', 'CheckIn_stars', 'CheckOut',
       'Cleanliness_stars', 'Communication_stars', 'Latitude',
       'Location_stars', 'Longitude', 'Price',
       ...
       'Smart lock', 'Smoke detector', 'Stair gates', 'Stove', 'TV',
       'Table corner guards', 'Washer', 'Waterfront', 'Wifi', 'Window guards'],
      dtype='object', length=107)

In [223]:
NewYorkData.head()

Accuracy_stars              CheckIn CheckIn_stars             CheckOut  \
0              5  2019-03-21 00:00:00             5  2019-03-26 00:00:00   
1            4.5  2019-03-21 00:00:00           4.5  2019-03-26 00:00:00   
2              5  2019-03-21 00:00:00             5  2019-03-26 00:00:00   
3              5  2019-03-21 00:00:00             5  2019-03-26 00:00:00   
4              5  2019-03-21 00:00:00             5  2019-03-26 00:00:00   

  Cleanliness_stars Communication_stars  Latitude Location_stars  Longitude  \
0                 5                   5  40.73321              5  -74.00242   
1                 4                 4.5  40.76580              5  -73.95812   
2                 5                   5  40.75972            4.5  -74.03827   
3                 5                   5  40.81386              5  -73.95936   
4                 5                   5  40.69284            4.5  -73.94418   

   Price      ...        Smart lock Smoke detector  Stair gates  Stove TV  \
0  244.0      ...                 0              1            0      1  1   
1   62.0      ...                 0              1            0      0  0   
2   44.0      ...                 0              1            0      0  0   
3   78.0      ...                 0              1            0      1  0   
4   80.0      ...                 0              1            0      0  0   

   Table corner guards  Washer  Waterfront  Wifi  Window guards  
0                    0       0           0     1              0  
1                    0       0           0     1              0  
2                    0       0           0     1              0  
3                    0       0           0     1              0  
4                    0       1           0     1              0  

[5 rows x 107 columns]

### Bangkok

In [224]:
print(len(BangkokData.columns))
BangkokData.columns

105


Index(['CheckIn', 'CheckOut', 'Price', 'Super_host', 'Latitude', 'Longitude',
       'Accuracy_stars', 'Communication_stars', 'Cleanliness_stars',
       'Location_stars',
       ...
       'Smart lock', 'Smoke detector', 'Stair gates', 'Stove', 'TV',
       'Table corner guards', 'Washer', 'Waterfront', 'Wifi', 'Window guards'],
      dtype='object', length=105)

In [225]:
BangkokData.head()

CheckIn         CheckOut  Price  Super_host  Latitude  Longitude  \
0  2019-03-12 0:00  2019-03-17 0:00   34.0           1  13.78165  100.57453   
1  2019-03-12 0:00  2019-03-17 0:00   48.0           0  13.69750  100.60708   
2  2019-03-12 0:00  2019-03-17 0:00   27.0           1  13.71252  100.46783   
3  2019-03-12 0:00  2019-03-17 0:00   35.0           0  13.72032  100.49623   
4  2019-03-12 0:00  2019-03-17 0:00   34.0           1  13.75738  100.56318   

  Accuracy_stars Communication_stars Cleanliness_stars Location_stars  \
0              5                   5               4.5              5   
1              0                   0                 0              0   
2              5                   5                 5              4   
3              5                   5               4.5              5   
4              0                   0                 0              0   

       ...       Smart lock Smoke detector  Stair gates  Stove TV  \
0      ...                0              1            0      1  1   
1      ...                0              1            0      0  1   
2      ...                0              1            0      0  1   
3      ...                0              1            0      1  1   
4      ...                0              0            0      0  1   

   Table corner guards  Washer  Waterfront  Wifi  Window guards  
0                    0       1           0     1              0  
1                    0       1           0     1              0  
2                    0       1           0     1              0  
3                    0       1           0     1              1  
4                    0       1           0     1              0  

[5 rows x 105 columns]

In [226]:
BangkokData.columns

Index(['CheckIn', 'CheckOut', 'Price', 'Super_host', 'Latitude', 'Longitude',
       'Accuracy_stars', 'Communication_stars', 'Cleanliness_stars',
       'Location_stars',
       ...
       'Smart lock', 'Smoke detector', 'Stair gates', 'Stove', 'TV',
       'Table corner guards', 'Washer', 'Waterfront', 'Wifi', 'Window guards'],
      dtype='object', length=105)

### Cairo

In [227]:
print(len(CairoData.columns))
CairoData.columns

106


Index(['CheckIn', 'CheckOut', 'Price', 'Super_host', 'Latitude', 'Longitude',
       'Accuracy_stars', 'Communication_stars', 'Cleanliness_stars',
       'Location_stars',
       ...
       'Smart lock', 'Smoke detector', 'Stair gates', 'Stove', 'TV',
       'Table corner guards', 'Washer', 'Waterfront', 'Wifi', 'Window guards'],
      dtype='object', length=106)

In [228]:
CairoData.head()

CheckIn         CheckOut  Price  Super_host  Latitude  Longitude  \
0  2019-03-14 0:00  2019-03-19 0:00   15.0           0  29.95961   31.25858   
1  2019-03-14 0:00  2019-03-19 0:00   12.0           0  30.05760   31.48779   
2  2019-03-14 0:00  2019-03-19 0:00   22.0           0  30.06085   31.49540   
3  2019-03-14 0:00  2019-03-19 0:00   19.0           0  30.06284   31.51743   
4  2019-03-14 0:00  2019-03-19 0:00   23.0           0  30.06263   31.48585   

  Accuracy_stars Communication_stars Cleanliness_stars Location_stars  \
0              0                   0                 0              0   
1              0                   0                 0              0   
2              4                   5               3.5            4.5   
3              5                   5                 5              5   
4              0                   0                 0              0   

       ...       Smart lock Smoke detector  Stair gates  Stove TV  \
0      ...                0              1            0      0  1   
1      ...                0              0            0      1  1   
2      ...                0              0            0      0  1   
3      ...                0              1            0      0  1   
4      ...                0              0            0      1  1   

   Table corner guards  Washer  Waterfront  Wifi  Window guards  
0                    0       1           0     1              0  
1                    0       1           0     1              1  
2                    0       1           0     1              0  
3                    0       1           0     1              0  
4                    0       1           0     1              0  

[5 rows x 106 columns]

### Geometry

In [6]:
NewYorkData_clean['geom']= NewYorkData_clean['Latitude'].map(str) +','+ NewYorkData_clean['Longitude'].map(str)
BangkokData_clean['geom']=  BangkokData_clean['Latitude'].map(str)+','+ BangkokData_clean['Longitude'].map(str) 
CairoData_clean['geom']= CairoData_clean['Latitude'].map(str)+','+  CairoData_clean['Longitude'].map(str) 

### Address string from geometry

In [27]:
from geopy.geocoders import Nominatim
geolocator = Nominatim(timeout=50,user_agent="Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:17.0) Gecko/20100101 Firefox/17.0")

In [240]:
NewYorkData_clean['address']=-1
BangkokData_clean['address']=-1
CairoData_clean['address']=-1

### NewYork

In [ ]:
for i,j in enumerate(NewYorkData_clean['geom']):
    NewYorkData_clean['address'][i]= geolocator.reverse(j).address 

In [15]:
NewYorkData_clean['address'].head()

0    204, West 4th Street, West Village, Manhattan,...
1    351, East 68th Street, Lenox Hill, Manhattan, ...
2    582, 12th Street, Union City, Hudson County, N...
3    3150, Broadway, Morningside Gardens, Morningsi...
4    186, Pulaski Street, Eleanor Roosevelt Houses,...
Name: address, dtype: object

In [8]:
geolocator.reverse(NewYorkData_clean['geom'][5],addressdetails=True).address

'Time Warner Center, 102, 8 Columbus Circle, Columbus Circle, Manhattan, Manhattan Community Board 5, New York County, NYC, New York, 10023, USA'

In [39]:
add='201, Sussex Street, Harrison, Hudson County, New Jersey, 07029, USA'
location = geolocator.geocode(add, addressdetails=True)
print(location.raw['address']['town'])

Harrison


In [9]:
NewYorkData_clean['partial_address']=-1

In [11]:
for i,j in enumerate(NewYorkData_clean['address']):
    NewYorkData_clean['partial_address'][i]= ','.join(j.split(',')[2:])
    

In [154]:
for i,j in enumerate(NewYorkData_clean['partial_address']):
    NewYorkData_clean['partial_address'][i]= ','.join(j)
    

In [27]:
add = NewYorkData_clean['partial_address'][3]

In [14]:
NewYorkData_clean['neighborhoodd']= -1

In [ ]:
for i,add in enumerate(NewYorkData_clean['address']):
    try:
        NewYorkData_clean['neighborhoodd'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['neighbourhood']
    except(AttributeError) as e:
        pass
    except(KeyError) as e:
        try:
            NewYorkData_clean['neighborhoodd'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['residential']
        except(KeyError) as e:
            try:
                NewYorkData_clean['neighborhoodd'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['residential']
            except(KeyError)as e:
                try:
                    NewYorkData_clean['neighborhoodd'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['suburb']
                except(KeyError)as e:
                    try:
                        NewYorkData_clean['neighborhoodd'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['city']
                    except(KeyError) as e:
                        try:
                             NewYorkData_clean['neighborhoodd'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['town']
                        except(KeyError) as e:
                            pass
                        

In [8]:
NewYorkData_clean['neighborhood'].head(10)

0                 West Village
1                   Lenox Hill
2                   Union City
3          Morningside Heights
4     Eleanor Roosevelt Houses
5              Columbus Circle
6                      TriBeCa
7              Upper West Side
8                       Harlem
9                  Fort Greene
Name: neighborhood, dtype: object

### Bangkok

In [12]:
BangkokData_clean['address'].head()

0    Mansion 7 Night Alive, ถนนรัชดาภิเษก, รัชดา, ก...
1    Bangkok College of Technology, Sukhumvit Soi 9...
2    ชุมชนเพชรเกษม 40, กรุงเทพมหานคร, เขตจอมทอง, กร...
3    อรคารไทยวีรวัฒน์, Saraphi 3, วัดพระยาไกร, กรุง...
4    Sap Thani, รัชดา, กรุงเทพมหานคร, เขตดินแดง, กร...
Name: address, dtype: object

In [10]:
BangkokData_clean['neighborhood']=-1

In [17]:
for i,add in enumerate(BangkokData_clean['address']):
    try:
        BangkokData_clean['neighborhood'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['neighbourhood']
    except(AttributeError) as e:
        pass
    except(KeyError) as e:
        try:
            BangkokData_clean['neighborhood'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['residential']
        except(KeyError) as e:
            try:
                BangkokData_clean['neighborhood'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['residential']
            except(KeyError)as e:
                try:
                    BangkokData_clean['neighborhood'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['suburb']
                except(KeyError)as e:
                    try:
                        BangkokData_clean['neighborhood'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['city']
                    except(KeyError) as e:
                        try:
                             BangkokData_clean['neighborhood'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['town']
                        except(KeyError) as e:
                            pass
                        

### Cairo

In [ ]:
for i,j in enumerate(CairoData_clean['geom']):
    CairoData_clean['address'][i]= geolocator.reverse(j).address  

In [48]:
for i,add in enumerate(CairoData_clean['address']):
    CairoData_clean['neighborhood'][i]= add.split(',')[1]

In [16]:
NewYorkData_clean['Reviewers_no'] = NewYorkData['Reviewers_no']
BangkokData_clean['Reviewers_no'] = BangkokData['Reviewers_no']
CairoData_clean['Reviewers_no'] = CairoData['Reviewers_no']


In [ ]:
import datetime
NewYorkData_clean['CheckIn'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in NewYorkData_clean['CheckIn']]
NewYorkData_clean['CheckOut'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in NewYorkData_clean['CheckOut']]


In [84]:
NewYorkData_clean['month'] = [i.month for i in NewYorkData_clean['CheckIn']]

In [ ]:
BangkokData_clean['CheckIn'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in BangkokData_clean['CheckIn']]
BangkokData_clean['CheckOut'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in BangkokData_clean['CheckOut']]
BangkokData_clean['month'] = [i.month for i in BangkokData_clean['CheckIn']]

In [75]:
CairoData_clean['CheckIn'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in CairoData_clean['CheckIn']]
CairoData_clean['CheckOut'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in CairoData_clean['CheckOut']]
CairoData_clean['month'] = [i.month for i in CairoData_clean['CheckIn']]

In [132]:
NewYorkData_clean['Title']= NewYorkData['Title']


In [133]:
BangkokData_clean['Title']= BangkokData['Title']


In [134]:
CairoData_clean['Title']= CairoData['Title']


In [135]:
NewYorkData_clean['Price'].min()

20.0

In [136]:
NewYorkData_clean['normalized_price']=NewYorkData_clean['Price']/NewYorkData_clean['Price'].min()

In [21]:
NewYorkData_clean['log_price']= NewYorkData_clean['Price'].apply(np.log)

In [11]:
NewYorkData_clean['CheckIn']=NewYorkData_clean['CheckIn'].apply(parser.parse)
NewYorkData_clean['CheckOut']=NewYorkData_clean['CheckOut'].apply(parser.parse)

In [13]:
NewYorkData_clean.isnull().sum().sort_values(ascending=False)

Bedrooms_int                  609
normalized_price               40
Price                          40
Baby bath                       0
Baby monitor                    0
Babysitter recommendations      0
Bathtub                         0
Beach essentials                0
Beachfront                      0
Bed linens                      0
Breakfast                       0
Building staff                  0
Cable TV                        0
Carbon monoxide detector        0
Changing table                  0
Air conditioning                0
Children’s books and toys       0
Children’s dinnerware           0
Cleaning before checkout        0
Coffee maker                    0
Cooking basics                  0
Crib                            0
Dishes and silverware           0
Dishwasher                      0
Dryer                           0
EV charger                      0
Elevator                        0
BBQ grill                       0
sofa                            0
Unnamed: 27   

In [ ]:
NewYorkData_clean.isemo().sum().sort_values(ascending=False)

In [14]:
NewYorkData_clean[NewYorkData_clean.Bedrooms_int.isnull().values]['Title'].head(30)

16                       Studio in the heart of Hoboken!
25                                   small room with key
29                                        Chelsea studio
50     Convenient & Comfortable Studio close to Manha...
58                                       118 Magnolia 2B
61               Beautiful studio apartment in Sunnyside
85           BEAUTIFUL BROWNSTONE STUDIO IN SOUTH HARLEM
86                                  Astoria, Queens, NYC
90     Studio apartment conveniently located in Manha...
91                 Union Square Private Sunny Dream Home
112      Super Luxury! Alcove Studio in heart of Chelsea
138                  Lrg Garden Studio w/Yard in 1843 RH
166                                      Sky Time Square
226               WeLive Wall Street -- Studio Apartment
259                           Private Room in Great Area
263                          Studio with Marble Bathroom
297    Large Condo in Bronx near Yankee stadium & subway
309               UNIQUE & COZY

So, we can see that most of the observations, with the number of bedrooms are assigned as NaN, are actually "Studio" So we can take advantage if this and use 1.5 as a key value for studio  

In [15]:
NewYorkData_clean['Bedrooms_int']=NewYorkData_clean['Bedrooms_int'].fillna(1.5)

In [16]:
NewYorkData_clean['Bedrooms_int'].unique()

array([1. , 1.5, 2. , 3. , 4. , 5. , 6. ])

In [19]:
NewYorkData_clean[NewYorkData_clean.Price.isnull().values].head()

Accuracy_stars    CheckIn  CheckIn_stars   CheckOut  Cleanliness_stars  \
296              0.0 2019-04-18            0.0 2019-04-23                0.0   
419              0.0 2019-05-16            0.0 2019-05-21                0.0   
495              5.0 2019-06-13            5.0 2019-06-18                5.0   
656              5.0 2019-07-11            5.0 2019-07-16                4.5   
1107             5.0 2019-10-03            5.0 2019-10-08                5.0   

      Communication_stars  Latitude  Location_stars  Longitude  Price  \
296                   0.0  40.72657             0.0  -73.98135    NaN   
419                   0.0  40.72160             0.0  -73.99713    NaN   
495                   4.5  40.71214             5.0  -73.99643    NaN   
656                   5.0  40.76236             5.0  -73.99179    NaN   
1107                  5.0  40.80239             5.0  -73.99437    NaN   

            ...         Stove  TV  Table corner guards  Washer  Waterfront  \
296         ...             0   0                    0       0           0   
419         ...             0   0                    0       0           0   
495         ...             1   0                    0       0           0   
656         ...             0   1                    0       1           0   
1107        ...             1   0                    0       1           0   

      Wifi  Window guards  Reviewers_no  \
296      1              0            -1   
419      1              0            -1   
495      1              0             6   
656      1              0            16   
1107     1              0            25   

                                                  Title  normalized_price  
296     Live like a local in the heart of East Village!               NaN  
419            Dreamy artist Apartment in SoHo / Nolita               NaN  
495   Sunny Room C, Downtown NYC, Chinatown, Two Bri...               NaN  
656          Luxury Room in Hell's Kitchen/Times Square               NaN  
1107         Cozy & bright apartment close to Manhattan               NaN  

[5 rows x 110 columns]

looks like most of the rows that have the "Price" missing,is ONE-bedroom size. So we can replace the empty/NaN values with the average price of ONE-bedroom homes

In [22]:
NewYorkData_clean[NewYorkData_clean.Bedrooms_int==1].Price.mean()

133.80853277835587

In [23]:
NewYorkData_clean['Price']=NewYorkData_clean['Price'].fillna(134)

In [144]:
NewYorkData_clean['normalized_price']=NewYorkData_clean['Price']/NewYorkData_clean['Price'].min()

In [145]:
NewYorkData_clean.isnull().sum().sort_values(ascending=False)

Unnamed: 114                  0
Air conditioning              0
Baby bath                     0
Baby monitor                  0
Babysitter recommendations    0
Bathtub                       0
Beach essentials              0
Beachfront                    0
Bed linens                    0
Breakfast                     0
Building staff                0
Cable TV                      0
Carbon monoxide detector      0
Changing table                0
Children’s books and toys     0
Children’s dinnerware         0
Cleaning before checkout      0
Coffee maker                  0
Cooking basics                0
Crib                          0
Dishes and silverware         0
Dishwasher                    0
Dryer                         0
EV charger                    0
Elevator                      0
Essentials                    0
Ethernet connection           0
BBQ grill                     0
Unnamed: 27                   0
Fire extinguisher             0
                             ..
Patio or

In [26]:
BangkokData_clean['CheckIn']= BangkokData_clean['CheckIn'].apply(parser.parse)
BangkokData_clean['CheckOut']= BangkokData_clean['CheckOut'].apply(parser.parse)

In [27]:
BangkokData_clean.isnull().sum().sort_values(ascending=False)

Price                         148
Title                           0
Carbon monoxide detector        0
BBQ grill                       0
Baby bath                       0
Baby monitor                    0
Babysitter recommendations      0
Bathtub                         0
Beach essentials                0
Beachfront                      0
Bed linens                      0
Breakfast                       0
Building staff                  0
Cable TV                        0
Changing table                  0
Unnamed: 26                     0
Children’s books and toys       0
Children’s dinnerware           0
Cleaning before checkout        0
Coffee maker                    0
Cooking basics                  0
Crib                            0
Dishes and silverware           0
Dishwasher                      0
Dryer                           0
EV charger                      0
Air conditioning                0
sofa                            0
Essentials                      0
Bathrooms_num_

In [31]:
BangkokData_clean[BangkokData_clean.Price.isnull().values]['Bedrooms_int'].head()

63     1
80     1
96    -1
165    1
230    1
Name: Bedrooms_int, dtype: int64

In [34]:

BangkokData_clean[BangkokData_clean.Bedrooms_int==1].Price.mean()

52.53104026845637

In [35]:
BangkokData_clean['Price']= BangkokData_clean['Price'].fillna(53)

In [146]:
BangkokData_clean['normalized_price']= BangkokData_clean['Price']/BangkokData_clean['Price'].min()


In [22]:
BangkokData_clean['log_price']= BangkokData_clean['Price'].apply(np.log)

In [147]:
BangkokData_clean.isnull().sum().sort_values(ascending=False)

month                         0
Changing table                0
Baby bath                     0
Baby monitor                  0
Babysitter recommendations    0
Bathtub                       0
Beach essentials              0
Beachfront                    0
Bed linens                    0
Breakfast                     0
Building staff                0
Cable TV                      0
Carbon monoxide detector      0
Children’s books and toys     0
Air conditioning              0
Children’s dinnerware         0
Cleaning before checkout      0
Coffee maker                  0
Cooking basics                0
Crib                          0
Dishes and silverware         0
Dishwasher                    0
Dryer                         0
EV charger                    0
Elevator                      0
Essentials                    0
BBQ grill                     0
Unnamed: 26                   0
geom                          0
Bathrooms_num_int             0
                             ..
Title   

In [148]:
CairoData_clean['normalized_price']= CairoData_clean['Price']/CairoData_clean['Price'].min()

In [23]:
CairoData_clean['log_price']= CairoData_clean['Price'].apply(np.log)

In [39]:
CairoData_clean['CheckIn']= CairoData_clean['CheckIn'].apply(parser.parse)
CairoData_clean['CheckOut']= CairoData_clean['CheckOut'].apply(parser.parse)


In [149]:
CairoData_clean.isnull().sum().sort_values(ascending=False)

month                         0
Changing table                0
Baby bath                     0
Baby monitor                  0
Babysitter recommendations    0
Bathtub                       0
Beach essentials              0
Beachfront                    0
Bed linens                    0
Breakfast                     0
Building staff                0
Cable TV                      0
Carbon monoxide detector      0
Children’s books and toys     0
Air conditioning              0
Children’s dinnerware         0
Cleaning before checkout      0
Coffee maker                  0
Cooking basics                0
Crib                          0
Dishes and silverware         0
Dishwasher                    0
Dryer                         0
EV charger                    0
Elevator                      0
Essentials                    0
BBQ grill                     0
Unnamed: 26                   0
Extra pillows and blankets    0
Bathrooms_num_int             0
                             ..
normaliz

In [30]:
CairoData_clean['neighborhood'] =-1

In [ ]:
for i,add in enumerate(CairoData_clean['address']):
    try:
        CairoData_clean['neighborhood'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['neighbourhood']
    except(AttributeError) as e:
        pass
    except(KeyError) as e:
        try:
            CairoData_clean['neighborhood'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['residential']
        except(KeyError) as e:
            try:
                CairoData_clean['neighborhood'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['residential']
            except(KeyError)as e:
                try:
                    CairoData_clean['neighborhood'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['suburb']
                except(KeyError)as e:
                    try:
                        CairoData_clean['neighborhood'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['city']
                    except(KeyError) as e:
                        try:
                             CairoData_clean['neighborhood'][i]= geolocator.geocode(add, addressdetails=True).raw['address']['town']
                        except(KeyError) as e:
                            pass
                        

In [38]:
from geopy.geocoders import Nominatim
geolocator = Nominatim(timeout=50,user_agent="Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:17.0) Gecko/20100101 Firefox/17.0")

In [31]:
geolocator.reverse(40.72576,-73.98309,addressdetails=True).address

'CV-110, Herbés, els Ports, Castelló / Castellón, Aragón, España'

In [25]:
geolocator.reverse(40.73321,-74.00242,addressdetails=True).address

'La Vega, Peñarroya de Tastavíns / Pena-roja de Tastavins, Matarraña, Teruel, Aragón, España'

In [94]:
NYDta = CairoData

In [95]:
NYDta['Group_ID'] = NYDta.groupby('geom').grouper.group_info[0]

In [96]:
NYDta.reset_index(inplace=True)

In [97]:
BangkokData.sort_index(by =['Group_ID','CheckIn'],inplace = True)

In [ ]:
NYDta[NYDta['Group_ID']==0]

In [ ]:
for i in NYDta['Group_ID']:
    for j,k in enumerate(NYDta['geom'][NYDta.Group_ID == i]):
        NYDta['address'][j]= geolocator.reverse(k,addressdetails=True, exactly_one=True).raw['address']

In [ ]:
for i in NYDta['Group_ID']:
    for j,k in enumerate(NYDta['geom'][NYDta.Group_ID == i]):
        print(NYDta['address'][j],k)

In [ ]:
for i in NYDta['Group_ID']:
    print(NYDta[['geom','neighborhood','Title']][NYDta['Group_ID'] == i])
    

## Day of the month

In [38]:
import datetime
NewYorkData_clean['CheckIn'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in NewYorkData['CheckIn']]
NewYorkData_clean['CheckOut'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in NewYorkData['CheckOut']]

In [41]:
NewYorkData_clean['day_of_month'] = [i.day for i in NewYorkData['CheckIn']]

In [45]:
BangkokData_clean['CheckIn'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in BangkokData['CheckIn']]
BangkokData_clean['CheckOut'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in BangkokData['CheckOut']]

In [46]:
BangkokData_clean['day_of_month'] = [i.day for i in BangkokData['CheckIn']]

In [69]:
CairoData_clean['CheckIn'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in CairoData_clean['CheckIn']]
CairoData_clean['CheckOut'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in CairoData_clean['CheckOut']]

In [70]:
CairoData_clean['day_of_month'] = [i.day for i in CairoData['CheckIn']]

## Week of the year

In [6]:
import datetime
NewYorkData_clean['CheckIn'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in NewYorkData_clean['CheckIn']]
NewYorkData_clean['CheckOut'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in NewYorkData_clean['CheckOut']]

In [11]:
NewYorkData_clean['Year_week']= week_of_year(NewYorkData_clean)

In [7]:
BangkokData_clean['CheckIn'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in BangkokData_clean['CheckIn']]
BangkokData_clean['CheckOut'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in BangkokData_clean['CheckOut']]

In [14]:
BangkokData_clean['Year_week']= week_of_year(BangkokData_clean)

In [8]:
CairoData_clean['CheckIn'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in CairoData_clean['CheckIn']]
CairoData_clean['CheckOut'] = [datetime.datetime.strptime(i,"%Y-%m-%d") for i in CairoData_clean['CheckOut']]

In [16]:
CairoData_clean['Year_week']= week_of_year(CairoData_clean)

## Number of weeks from the point of collecting data to the booking time

Since the date of collecting the data slightly differs among the three cities, we need to extract the "weeks_before_booking" variable separately for the three cities. 

## New York

In [33]:
NewYorkData_clean['weeks_before_booking'] = None

The day of collecting data "Data Scraping"

In [30]:
NewYorkData_clean['CheckIn'][1]

Timestamp('2019-03-07 00:00:00')

which week of the year that day?!

In [31]:
NewYorkData_clean['CheckIn'][1].isocalendar()[1]

10

In [35]:
for i, j in enumerate(NewYorkData_clean['CheckIn']):
    if j.month >= 3:
        NewYorkData_clean['weeks_before_booking'][i]= j.isocalendar()[1] - 10
    else:
        NewYorkData_clean['weeks_before_booking'][i]= j.isocalendar()[1] + 42
        

## Bangkok

In [38]:
BangkokData_clean['weeks_before_booking'] = None

The day of collecting data "Data Scraping"

In [42]:
BangkokData_clean['CheckIn'][22]

Timestamp('2019-03-12 00:00:00')

which week of the year that day?!

In [43]:
BangkokData_clean['CheckIn'][22].isocalendar()[1]

11

In [42]:
for i, j in enumerate(BangkokData_clean['CheckIn']):
    if j.month >= 3:
        BangkokData_clean['weeks_before_booking'][i]= j.isocalendar()[1] - 11
    else:
        BangkokData_clean['weeks_before_booking'][i]= j.isocalendar()[1] + 42

## Cairo

In [44]:
CairoData_clean.head()

index    CheckIn   CheckOut  Price  Super_host  Latitude  Longitude  \
0   3201 2019-08-01 2019-08-06   39.0           1  30.05867   31.44884   
1    908 2019-04-11 2019-04-16   27.0           0  29.96925   31.26652   
2   4605 2019-10-10 2019-10-15   51.0           1  29.95475   31.26336   
3   5984 2019-12-05 2019-12-10   21.0           0  30.12274   31.60838   
4   3680 2019-08-29 2019-09-03   33.0           1  30.04762   31.23734   

   Accuracy_stars  Communication_stars  Cleanliness_stars    ...      \
0             0.0                  0.0                0.0    ...       
1             0.0                  0.0                0.0    ...       
2             4.0                  4.5                4.5    ...       
3             0.0                  0.0                0.0    ...       
4             5.0                  5.0                5.0    ...       

   AvgPrice_neighborhood  AvgNormPrice_neighborhood  Group_ID  \
0                     54                          4        81   
1                     53                          4        21   
2                     54                          4        81   
3                     53                          4        43   
4                     52                          4        66   

   avg_rating_stars  price_stars_corr_per_nhood  price_stars_corr_Zvalue  \
0          0.000000                   -0.015865                -0.015867   
1          0.000000                    0.045170                 0.045200   
2          4.416667                   -0.015865                -0.015867   
3          0.000000                   -0.018695                -0.018697   
4          5.000000                    0.236721                 0.241298   

   avg_price_stars_corr  Amenities_length  day_of_month  Year_week  
0             -0.015865                23             1         31  
1              0.045170                24            11         15  
2             -0.015865                26            10         41  
3             -0.018695                18             5         49  
4              0.236721                28            29         35  

[5 rows x 124 columns]

In [58]:
CairoData_clean['weeks_before_booking'] = None

The day of collecting data "Data Scraping"

In [56]:
CairoData_clean['CheckIn'][20]

Timestamp('2019-03-14 00:00:00')

which week of the year that day?!

In [57]:
CairoData_clean['CheckIn'][20].isocalendar()[1]

11

In [59]:
for i, j in enumerate(CairoData_clean['CheckIn']):
    if j.month >= 3:
        CairoData_clean['weeks_before_booking'][i]= j.isocalendar()[1] - 11
    else:
        CairoData_clean['weeks_before_booking'][i]= j.isocalendar()[1] + 42

In [60]:
CairoData_clean['weeks_before_booking'].unique()

array([20, 4, 30, 38, 24, 12, 18, 32, 49, 6, 14, 47, 34, 43, 0, 10, 26,
       45, 22, 2, 36, 51, 28, 40, 8, 16], dtype=object)

## Number of listings in the week

In [66]:
NewYorkData_clean['ListingsNo_in_weekth']= None
for i, j in enumerate(NewYorkData_clean['Year_week']):
    NewYorkData_clean['ListingsNo_in_weekth'][i] = sum(NewYorkData_clean['Year_week']==j)

In [68]:
BangkokData_clean['ListingsNo_in_weekth']= None
for i, j in enumerate(BangkokData_clean['Year_week']):
    BangkokData_clean['ListingsNo_in_weekth'][i] = sum(BangkokData_clean['Year_week']==j)

In [70]:
CairoData_clean['ListingsNo_in_weekth']= None
for i, j in enumerate(CairoData_clean['Year_week']):
    CairoData_clean['ListingsNo_in_weekth'][i] = sum(CairoData_clean['Year_week']==j)

## Saving the data

In [19]:
NewYorkData_clean.to_csv(path_or_buf= Data_dir/'NewYorkData_clean.csv',columns= NewYorkData_clean.columns,index=False)
BangkokData_clean.to_csv(path_or_buf= Data_dir/'BangkokData_clean.csv',columns= BangkokData_clean.columns,index=False,encoding='utf-8-sig')
CairoData_clean.to_csv(path_or_buf= Data_dir/'CairoData_clean.csv',columns= CairoData_clean.columns,index=False,encoding='utf-8-sig')